File: games/BastionTD/src_python/economy.py
"""
economy.py - Gold, lives, cost validation, and wave-clear bonus for Bastion TD.
"""
from settings import START_GOLD, START_LIVES, WAVE_CLEAR_BONUS


class Economy:
    """Tracks the player's gold and lives."""

    def __init__(self):
        self.gold: int = START_GOLD
        self.lives: int = START_LIVES
        self.total_earned: int = 0

    def can_afford(self, cost: int) -> bool:
        """Return True if the player has enough gold."""
        return self.gold >= cost

    def spend(self, cost: int) -> None:
        """Subtract gold (caller must check can_afford first)."""
        self.gold -= cost

    def earn(self, amount: int) -> None:
        """Add gold and track total earned."""
        self.gold += amount
        self.total_earned += amount

    def lose_lives(self, amount: int) -> None:
        """Subtract lives."""
        self.lives -= amount

    def is_game_over(self) -> bool:
        """True if lives have reached zero or below."""
        return self.lives <= 0

    def wave_clear_bonus(self) -> None:
        """Award the wave-clear bonus gold."""
        self.earn(WAVE_CLEAR_BONUS)

File: games/BastionTD/src_cpp/core/economy.cpp
#include "core/economy.h"

#include <cmath>

void Economy::reset() {
    gold = cfg::START_GOLD;
    lives = cfg::START_LIVES;
    had_leak_this_wave = false;
}

bool Economy::can_afford(int cost) const { return gold >= cost; }

void Economy::spend(int cost) { gold -= cost; }

void Economy::earn(int amount) { gold += amount; }

void Economy::lose_lives(int amount) {
    lives -= amount;
    had_leak_this_wave = true;
    if (lives < 0) {
        lives = 0;
    }
}

bool Economy::is_game_over() const { return lives <= 0; }

void Economy::wave_clear_bonus() {
    if (!had_leak_this_wave) {
        gold += cfg::WAVE_CLEAR_BONUS;
    }
    had_leak_this_wave = false;
}

int Economy::fleet_upgrade_cost(int tower_count, int per_tower_cost) {
    const float raw = static_cast<float>(tower_count * per_tower_cost);
    return static_cast<int>(std::ceil(raw * cfg::FLEET_UPGRADE_PREMIUM));
}
